storage/innobase/handler/ha_innodb.cc

```diff
@@ -3995,12 +3995,13 @@ bool innobase_fix_tablespaces_empty_uuid() {
     return (false);
   }
 
-  log_enable_encryption_if_set();
-
   /* We only need to handle the case when an encrypted tablespace
   is created at startup. If it is 0, there is no encrypted tablespace,
   If it is > 1, it means we already have fixed the UUID */
   if (Encryption::s_master_key_id != 1) {
+    /* keyring_key based redo log encryption has to be  initialized even in
```

Comment:
This comment is not helpful. The question "why" remains.

Reply:
Okay, I'll expand that comment. Basically this is the only place where we initialize the redo log encryption now, and keyring_key has to work even when there isn't anything encrypted with master_key - this was the reason why the call originally was before this condition.

Reply:
Updated the comment in the 5.7 PR - if it's good, I'll update this version too.

Reply:
I think there is problem with how(or where) we are enabling the encryption of redo. I think it should be done at startup code (before innodb_inited = true is set).
I missed this point in review of PS-4985.
Fix empty uuid is doing more than it is supposed to do and also becoming hackish.

So here is my suggestion:
1. Enable redo log encryption at startup code (before innodb_inited is set to true). Similar to how temp is encrypted at startup.
2. Remove log_enable_encryption_if_set() from innobase_fix_tablespaces_empty_uuid().
3. Keep the log_rotate_encryption() rotate calls in  innobase_fix_tablespaces_empty_uuid().
4. call log_enable_encryption_if_set() from the SET variable update functions. (For example, see innodb_temp_tablespace_encryption_update())

Reply:
Updated with (4), and backported some testcase fixes already done in 8.0

Reply:
Well, before this patch too, log_enable_encryption_if_set() is called when uuid is empty ?

Similar problems exists with MK too. before keyring is intialized, there is need for key.

So MK uses  "default key" during init which is fixed by fix_empty_..(). Cant you use something similar?

Reply:
I am using something similar, yes. But that doesn't change the base problem: log_enable_encryption has to be called after the uuid exists and innodb is inited, without that, it won't perform a full initialization for RK.

I could move some of the initialization logic to the rotation function, but that would be also a hack, as it's initialization, not rotation.

Also, this function currently exists early if there's no master key, which means it still wouldn't call rotate for RK redo logs, if nothing else is encrypted - a "hack" calling rotate in this condition still would be required.

Reply:
"Well, before this patch too, log_enable_encryption_if_set() is called when uuid is empty " - before the original patch (as in the original update code) it is called periodically every time the master loop is executed... :)

Reply:
Ok, if log_enable_encyrption() has to be called even after using something like default master key, then there is no good way to solve this.

Do you need the extra call only when s_uuid == 0? Otherwise we are calling log_enable_encryption() two times on every startup?

Reply:
For both RK and MK, I want log_enable_encryption() to be called only once (code can be present at multiple places), even though it wouldn't hurt performance. Is it possible to achieve without making it too ugly?

Reply:
Another question ( I think you added this). WHy is  log_rotate_encryption() present twice?

/* Rotate log tablespace */
  bool failure1 = !log_rotate_encryption();

  bool failure2 =
      !fil_encryption_rotate_global(space_ids) || !log_rotate_encryption();

Reply:
Uh, that's a merge mistake in 8.0, I'll also fix that.

Reply:
5.7 PR updated with an improved condition.

Reply:
will check, thank you

Reply:
looks good! and double log_rotation() problem exists in 5.7 too